core/src/main/java/io/grpc/internal/ClientCallImpl.java

```diff
@@ -143,9 +154,24 @@ static void prepareHeaders(Metadata headers, CallOptions callOptions, String use
   }
 
   @Override
-  public void start(Listener<RespT> observer, Metadata headers) {
+  public void start(final Listener<RespT> observer, Metadata headers) {
     Preconditions.checkState(stream == null, "Already started");
 
+    if (context.isCancelled()) {
+      // Context is already cancelled, cancel immediately and notify the observer of cancellation
+      // via callback on the executor
+      cancel();
+      callExecutor.execute(new ContextRunnable(context) {
+        @Override
+        public void runInContext() {
+          // Must call onHeaders per API contract
```

Comment:
I think that's a mistake in the docs ([left over](https://github.com/grpc/grpc-java/commit/e4bd1c7d69e62fa87250cc4bf4bd894bb3d51359) from before some old semantic changes to the protocol). I am very certain we will close without any onHeaders in error conditions, and I would believe we would want to continue. I would claim "onHeaders is always called before onPayload".